### PR TITLE
feat: add redis support in risedev

### PR DESCRIPTION
**Pull request by tabVersion**
Signed-off-by: tabVersion <tabvision@bupt.icu>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

as title, you can run `./risedev ci-start ci-redis` to start a cluster along with a redis instance

note that we start simply a standalone redis instead of a redis cluster 

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)

## Checklist

~~- [ ] I have written necessary rustdoc comments~~
~~- [ ] I have added necessary unit tests and integration tests~~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

#4240

### CODE REVIEW

File: src/risedevtool/redis.toml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+dependencies = ["prepare"]
+condition = { env_set = [ "ENABLE_REDIS" ] }
+description = "Download and extract Redis"
+script = '''
+#!/bin/bash
+set -e
+if [ -d "${PREFIX_BIN}/redis" ]; then
+    exit 0
+fi
+echo "Kafka not found, download ${REDIS_RELEASE}"
```
Comment (skyzh) [inline src/risedevtool/redis.toml:20]: Redis not found

File: src/risedevtool/redis.toml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+curl -fL -o "${REDIS_DOWNLOAD_PATH}" "${REDIS_DOWNLOAD_URL}"
+tar -xvzf "${REDIS_DOWNLOAD_PATH}" -C "${PREFIX_TMP}"
+mv "${PREFIX_TMP}/${REDIS_RELEASE}" "${PREFIX_BIN}/redis"
+rm ${REDIS_DOWNLOAD_PATH}
+cd "${PREFIX_BIN}/redis" && make -j all
```
Comment (skyzh) [inline src/risedevtool/redis.toml:25]: Don’t redis have pre-built binaries? Looks unacceptable for me to compile it from source.

File: src/risedevtool/src/bin/risedev-config.rs [hunk 3]
```diff
@@ -68,6 +68,7 @@ pub enum Components {
     Release,
     AllInOne,
     Sanitizer,
+    Redis,
```
Comment (skyzh) [inline src/risedevtool/src/bin/risedev-config.rs:71]: Put it along with other components enum

File: src/risedevtool/src/task/redis_service.rs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    }
+}
+
+impl Task for RedisService {
+    fn execute(&mut self, ctx: &mut ExecuteContext<impl Write>) -> Result<()> {
+        ctx.service(self);
+        ctx.pb.set_message("starting");
+        let path = self.redis_path()?;
+        if !path.exists() {
+            return Err(anyhow!("Kafka binary not found in {:?}\nDid you enable redis feature in `./risedev configure`?", path));
```
Comment (skyzh) [inline src/risedevtool/src/task/redis_service.rs:52]: Redis binary not found

**Comment by tabVersion** [comment]
I ran into a problem that ci-env cannot start correctly. I got the following log from frontend. 

```
2022-07-27T16:01:20.560830Z  INFO risingwave_rt: parking lot deadlock detection enabled
2022-07-27T16:01:20.573806Z  INFO risingwave_frontend::session: Starting frontend node with config FrontendConfig { server: ServerConfig { heartbeat_interval_ms: 1000 } }
2022-07-27T16:01:20.618883Z ERROR pgwire::pg_protocol: Error: ReadMsgError: unexpected end of file.
```

for tmux log, some window may exit abnormally at random places. 

```
> tmux -V
tmux 3.3a
(tmux exited with Some(0))
---
> tmux list-windows \
  -t \
  risedev \
  -F \
  #{pane_id} #{pane_pid} #{window_name}
no server running on /private/tmp/tmux-501/default
(tmux exited with Some(1))
---
> tmux kill-session \
  -t \
  risedev
no server running on /private/tmp/tmux-501/default
(tmux exited with Some(1))
---
```

**Review by skyzh — COMMENTED** [review]
Rest LGTM

**Comment by skyzh** [comment]
Also you’ll need to retry for Redis check. Otherwise it will fail immediately.

File: src/risedevtool/redis.toml (hunk 2 above)
Comment (tabVersion) [reply src/risedevtool/redis.toml:25]: I am also surprised but it does not have one. The compile takes 10+s on my machine and looks like it is acceptable.

Comment (skyzh) [reply src/risedevtool/redis.toml:25]: Does it need any dependency to compile?

File: src/workspace-hack/Cargo.toml [hunk 5]
```diff
@@ -136,9 +137,8 @@ scopeguard = { version = "1", features = ["use_std"] }
 serde = { version = "1", features = ["alloc", "derive", "serde_derive", "std"] }
 serde_json = { version = "1", features = ["raw_value", "std"] }
 smallvec = { version = "1", default-features = false, features = ["serde", "union", "write"] }
-socket2 = { version = "0.4", default-features = false, features = ["all"] }
 syn = { version = "1", features = ["clone-impls", "derive", "extra-traits", "full", "parsing", "printing", "proc-macro", "quote", "visit", "visit-mut"] }
-tokio = { version = "1", features = ["bytes", "fs", "io-std", "io-util", "libc", "macros", "memchr", "mio", "net", "num_cpus", "once_cell", "process", "rt", "rt-multi-thread", "signal", "signal-hook-registry", "socket2", "stats", "sync", "time", "tokio-macros", "tracing"] }
+tokio = { version = "1", features = ["bytes", "fs", "full", "io-std", "io-util", "libc", "macros", "memchr", "mio", "net", "num_cpus", "once_cell", "parking_lot", "process", "rt", "rt-multi-thread", "signal", "signal-hook-registry", "socket2", "stats", "sync", "time", "tokio-macros", "tracing"] }
```
Comment (skyzh) [inline src/workspace-hack/Cargo.toml:141]: Some crate enables `parking_lot` feature in tokio. Not sure if this would affect the performance very much.

**Review by skyzh — APPROVED** [review]
Rest LGTM

File: src/workspace-hack/Cargo.toml (hunk 5 above)
Comment (tabVersion) [reply src/workspace-hack/Cargo.toml:141]: looks there already have these two features in the toml

File: src/risedevtool/redis.toml (hunk 2 above)
Comment (tabVersion) [reply src/risedevtool/redis.toml:25]: No

**Comment by tabVersion** [comment]
let's hold the pr and wait for https://github.com/aembke/fred.rs/pull/51 merge. 
It can help not introduce useless features in tokio

**Pull request merged**